[Merged by Bors] - feat: definition and properties of characteristic functions

JakobStiefel opened the pull request:
define characteristic functions. Show that characteristic functions separate finite measures. 

---

- [x] depends on: #19780
- [x] depends on: #19781
- [x] depends on: #19782
- [x] depends on: #20604

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

RemyDegenne commented:
At some point I wrote code for the star-subalgebra of exponential polynomials (which is the same as what you do for `probChar_starSubalgebra`, but I did not take a general `AddChar`). I took some time today to generalize my code to match the generality you assumed, for the sake of comparison with yours. It's here: https://github.com/RemyDegenne/CLT/blob/tight/Clt/ExpPoly.lean (in the section `PR` at the beginning of the file; disregard the bad names for the defs).

The way I did it was to first define a monoid homomorphism `Multiplicative W →* (V →ᵇ ℂ)` from `probChar`. The monoid homomorphism can be lifted to a homomorphism of algebras from `AddMonoidAlgebra ℝ W` to `V →ᵇ ℂ` using `AddMonoidAlgebra.lift`. The subalgebra we want is the range of this homomorphism. The “star” part can finally be checked easily. The proofs of separation are copied (and minimally adapted) from your code.
It looks like that approach leads to simpler proofs compared to what you did, and it avoids the big proof of boundedness you have in `probChar_starSubalgebra_bounded`. What do you think?

JakobStiefel commented:
> At some point I wrote code for the star-subalgebra of exponential polynomials (which is the same as what you do for `probChar_starSubalgebra`, but I did not take a general `AddChar`). I took some time today to generalize my code to match the generality you assumed, for the sake of comparison with yours. It's here: https://github.com/RemyDegenne/CLT/blob/tight/Clt/ExpPoly.lean (in the section `PR` at the beginning of the file; disregard the bad names for the defs).
> 
> The way I did it was to first define a monoid homomorphism `Multiplicative W →* (V →ᵇ ℂ)` from `probChar`. The monoid homomorphism can be lifted to a homomorphism of algebras from `AddMonoidAlgebra ℝ W` to `V →ᵇ ℂ` using `AddMonoidAlgebra.lift`. The subalgebra we want is the range of this homomorphism. The “star” part can finally be checked easily. The proofs of separation are copied (and minimally adapted) from your code. It looks like that approach leads to simpler proofs compared to what you did, and it avoids the big proof of boundedness you have in `probChar_starSubalgebra_bounded`. What do you think?

Nice! Compared to your code, I now defined `expPoly` as subalgebra over `ℂ` instead of `ℝ`, which makes it easier to apply `ext_of_forall_mem_subalgbra_integral_eq_of_pseudoEMetric_complete_countable`. 

As an example, this can be used to show that moment generating functions are separating. I added some code on this, what do you think?

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+-/
+theorem MeasureTheory.ext_of_charFun_eq (he : Continuous e) (he' : e ≠ 1)
+    (hL' : ∀ v ≠ 0, L v ≠ 0) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
+    (P P' : Measure V) [IsFiniteMeasure P] [IsFiniteMeasure P'] :
+    (∀ w, ∫ v, probChar he hL w v ∂P = ∫ v, probChar he hL w v ∂P') → P = P' := by
+  intro h
```
```suggestion
    (P P' : Measure V) [IsFiniteMeasure P] [IsFiniteMeasure P']
    (h : ∀ w, ∫ v, probChar he hL w v ∂P = ∫ v, probChar he hL w v ∂P') :
    P = P' := by
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+  _ ≠ 1 := ha
+
+/-- Monoid homomorphism mapping `w` to `fun v ↦ e (L v (Multiplicative.toAdd w))`. -/
+def expInnerMulI (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```
The name `expInnerMulI` made sense for an old version of my code where `e` was the exponential of an inner product multiplied with `I`, but it does not really make sense here for a general `e`.
Same remark for `expInnerMulIₐ` or even `expPoly`.

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+
+open MeasureTheory Filter BoundedContinuousFunction Complex
+
+section ProbChar
```
Please put the content of this file in the `MeasureTheory` namespace.

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+import Mathlib.MeasureTheory.Measure.FiniteMeasureExt
+/-!
+
+# Characteristic Function of a Finite Measure
+
+This file defines the characteristic function of a `FiniteMeasure P` on a topological vector space
+`V`.
+
+## Probability Character
+
+We first define
+
+`probChar _ _ w : V →ᵇ ℂ := fun (v : V) ↦ e (L v w)`,
+
+where `e` is a continuous additive character and `L : V →ₗ[ℝ] W →ₗ[ℝ] ℝ` is a bilinear map. We
+then define `expPoly`, the star subalgebra of `ℂ`-linear combinations of elements of
+`{probChar _ _ w : w ∈ W}`. We show that `expPoly` separates points in `V`, see
+`expPoly_separatesPoints`.
+
+## Characterictic Function
+
+The characteristic function of a finite measure `P` on `V` is the mapping
+
+`fun w => ∫ v, e (-L v w) ∂P = ∫ v, probChar w ∂P`
+
+We show:
+- `MeasureTheory.ext_of_charFun_eq`: If the characteristic functions of two finite measures `P`
+  and `P'` are equal, then `P = P'`. In other words, characteristic functions separate finite
+  measures.
```
Please see https://leanprover-community.github.io/contribute/doc.html for the formatting rules of file docstrings. In particular, we have standard sections like "## Main definitions" and "## Main statements".
If you type "module" in the file and look at auto-complete suggestions, you will get an option to generate a template module docstring with the correct sections (delete the sections that don't make sense).

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+  _ ≠ 1 := ha
+
+/-- Monoid homomorphism mapping `w` to `fun v ↦ e (L v (Multiplicative.toAdd w))`. -/
+def probChar_monoidHom (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```
Definition names don't use `_`:
```suggestion
def probCharMonoidHom (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+
+/-- Algebra homomorphism mapping `w` to `fun v ↦ e (L v (Multiplicative.toAdd w))`. -/
+noncomputable
+def probChar_AlgHom (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```
```suggestion
def probCharAlgHom (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+
+/-- The star-subalgebra of exponential polynomials. -/
+noncomputable
+def probChar_Poly (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```
```suggestion
def probCharPoly (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2) :
```

RemyDegenne commented on Mathlib/Probability/Moments/ComplexMGF.lean:
```diff
@@ -297,4 +298,42 @@ lemma eqOn_complexMGF_of_mgf [IsProbabilityMeasure μ]
+
+variable {Ω' : Type*} {mΩ' : MeasurableSpace Ω'} {Y : Ω' → ℝ} {μ' : Measure Ω'}
+
+/-- Exponential map onto the circle, defined as additive character -/
+noncomputable
+def probFourierChar : AddChar ℝ Circle where
+  toFun z := Circle.exp (z)
+  map_zero_eq_one' := by simp only; rw [Circle.exp_zero]
+  map_add_eq_mul' x y := by simp only; rw [Circle.exp_add]
```
You could rename that `Circle.expAddChar` and move it to the Analysis/Complex/Circle file, since it's very similar to `Circle.expHom`. Put it just before the `variable` line that you added?

The next lemma `probFourierChar_ne_one` (to be renamed `expAddChar_ne_one`) can't be moved there though, since that file does not import enough to know about `Real.pi` (and we should probably not add too many imports there).

RemyDegenne commented on Mathlib/Probability/Moments/ComplexMGF.lean:
```diff
@@ -297,4 +298,42 @@ lemma eqOn_complexMGF_of_mgf [IsProbabilityMeasure μ]
+    rw [Circle.coe_exp Real.pi, Complex.exp_pi_mul_I] at heq
+    norm_num at heq
+
+theorem MeasureTheory.ext_of_forall_complexMGF_eq [IsFiniteMeasure μ]
+    [IsFiniteMeasure μ'] (hX : AEMeasurable X μ) (hY : AEMeasurable Y μ')
+    (h : ∀ z, complexMGF X μ z = complexMGF Y μ' z) :
+    μ.map X = μ'.map Y := by
+  have inner_ne_zero (x : ℝ) (h : x ≠ 0) : bilinFormOfRealInner x ≠ 0 :=
+    DFunLike.ne_iff.mpr ⟨x, inner_self_ne_zero.mpr h⟩
+  apply MeasureTheory.ext_of_charFun_eq Circle.exp.continuous probFourierChar_ne_one
+    inner_ne_zero continuous_inner (μ.map X) (μ'.map Y)
+  intro w
+  specialize h (Multiplicative.toAdd w * I)
+  simp [complexMGF] at h
+  simp_rw [mul_assoc, mul_comm I, ← mul_assoc] at h
+  simp [probChar_apply, probFourierChar]
+  rwa [integral_map hX (AEMeasurable.aestronglyMeasurable <| by fun_prop),
+    integral_map hY (AEMeasurable.aestronglyMeasurable <| by fun_prop)]
```
Several suggestions:
- use `_root_` to actually put it in the `MeasureTheory` namespace, not in `ProbabilityTheory.MeasureTheory`. Actually I think it should be in `MeasureTheory.Measure`.
- remove the forall from the hypothesis of equality of `complexMGF`. Just use equality of functions. Then add a `funext_iff` in the proof to get the forall form.
- There are two nonterminal `simp`: these should be `simp only`
```suggestion
theorem _root_.MeasureTheory.Measure.ext_of_complexMGF_eq [IsFiniteMeasure μ]
    [IsFiniteMeasure μ'] (hX : AEMeasurable X μ) (hY : AEMeasurable Y μ')
    (h : complexMGF X μ = complexMGF Y μ') :
    μ.map X = μ'.map Y := by
  have inner_ne_zero (x : ℝ) (h : x ≠ 0) : bilinFormOfRealInner x ≠ 0 :=
    DFunLike.ne_iff.mpr ⟨x, inner_self_ne_zero.mpr h⟩
  apply MeasureTheory.ext_of_charFun_eq Circle.exp.continuous expAddChar_ne_one
    inner_ne_zero continuous_inner (μ.map X) (μ'.map Y)
  intro w
  rw [funext_iff] at h
  specialize h (Multiplicative.toAdd w * I)
  simp_rw [complexMGF, mul_assoc, mul_comm I, ← mul_assoc] at h
  simp only [Circle.expAddChar, probChar_apply, bilinFormOfRealInner_apply_apply,
    RCLike.inner_apply, conj_trivial, AddChar.coe_mk, Circle.coe_exp, ofReal_mul]
  rwa [integral_map hX (AEMeasurable.aestronglyMeasurable <| by fun_prop),
    integral_map hY (AEMeasurable.aestronglyMeasurable <| by fun_prop)]
```

RemyDegenne commented on Mathlib/Probability/Moments/ComplexMGF.lean:
```diff
@@ -297,4 +298,42 @@ lemma eqOn_complexMGF_of_mgf [IsProbabilityMeasure μ]
+  map_zero_eq_one' := by simp only; rw [Circle.exp_zero]
+  map_add_eq_mul' x y := by simp only; rw [Circle.exp_add]
+
+lemma probFourierChar_ne_one : probFourierChar ≠ 1 := by
+    rw [DFunLike.ne_iff]
+    use Real.pi
+    simp only [probFourierChar, AddChar.coe_mk, AddChar.one_apply]
+    intro h
+    have heq := congrArg Subtype.val h
+    rw [Circle.coe_exp Real.pi, Complex.exp_pi_mul_I] at heq
+    norm_num at heq
```
```suggestion
lemma probFourierChar_ne_one : probFourierChar ≠ 1 := by
  rw [DFunLike.ne_iff]
  use Real.pi
  simp only [probFourierChar, AddChar.coe_mk, AddChar.one_apply]
  intro h
  have heq := congrArg Subtype.val h
  rw [Circle.coe_exp Real.pi, Complex.exp_pi_mul_I] at heq
  norm_num at heq
```

RemyDegenne reviewed (commented):
Here is a review of the ComplexMGF file. I'll make a more complete review later.
It's nice that you could add that extensionality result for complexMGF!

Perhaps the following lemma could be added at the end:
```lean
lemma _root_.MeasureTheory.Measure.ext_of_complexMGF_id_eq
    {μ μ' : Measure ℝ} [IsFiniteMeasure μ] [IsFiniteMeasure μ']
    (h : complexMGF id μ = complexMGF id μ') :
    μ = μ' := by
  simpa using Measure.ext_of_complexMGF_eq aemeasurable_id aemeasurable_id h
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+  · simp [Finsupp.single_apply_ne_zero]
+  · simp
+
+lemma probCharPoly_separatesPoints (he : Continuous e) (he' : e ≠ 1)
```
```suggestion
lemma separatesPoints_probCharPoly (he : Continuous e) (he' : e ≠ 1)
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+    (P P' : Measure V) [IsFiniteMeasure P] [IsFiniteMeasure P']
+    (h : ∀ w, ∫ v, probChar he hL w v ∂P = ∫ v, probChar he hL w v ∂P') :
+    P = P' := by
+  apply ext_of_forall_mem_subalgebra_integral_eq_of_pseudoEMetric_complete_countable
+      (probCharPoly_separatesPoints he he' hL hL')
+  intro g hg
+  simp [StarSubalgebra.mem_map, mem_probCharPoly] at hg
```
non-terminal simp.
```suggestion
  apply ext_of_forall_mem_subalgebra_integral_eq_of_pseudoEMetric_complete_countable
      (probCharPoly_separatesPoints he he' hL hL') fun g hg ↦ ?_
  simp only [mem_probCharPoly] at hg
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+  obtain ⟨w, hw⟩ := hg
+  rw [hw]
+  have hsum (P : Measure V) [IsFiniteMeasure P] :
+      ∫ v, ∑ a ∈ w.support, w a * ↑(e ((L v) a)) ∂P
+      = ∑ a ∈ w.support, ∫ v, w a * ↑(e ((L v) a)) ∂P :=
```
It's more readable if we remove coercions and unnecessary parentheses:
```suggestion
      ∫ v, ∑ a ∈ w.support, w a * e (L v a) ∂P = ∑ a ∈ w.support, ∫ v, w a * e (L v a) ∂P :=
```

RemyDegenne commented on Mathlib/MeasureTheory/Integral/FiniteMeasureCharFun.lean:
```diff
@@ -0,0 +1,223 @@
+    integral_finset_sum w.support
+      fun a ha => Integrable.const_mul (integrable P (probChar he hL a)) _
+  rw [hsum P, hsum P']
+  apply Finset.sum_congr rfl
+  intro i _
+  simp only [smul_eq_mul, MeasureTheory.integral_mul_left, mul_eq_mul_left_iff]
+  left
+  exact h i
```
A very minor shortening of the code. Feel free to ignore this comment.
```suggestion
  apply Finset.sum_congr rfl fun i _ ↦ ?_
  simp only [smul_eq_mul, MeasureTheory.integral_mul_left, mul_eq_mul_left_iff]
  exact Or.inl (h i)
```

RemyDegenne reviewed (commented):
The name and the placement of the file `FiniteMeasureCharFun` are a bit strange to me. Why is it in the Integral folder? Since it is defined from a measure, I would have put it in the Measure folder. And the name of the file could perhaps be `CharacteristicFunction`?

You don't actually add a definition for the characteristic function. Perhaps it would be worth adding a definition `charFun P hL w` to refer to the integral of `probChar` for the standard choice of  `e = probFourierChar` (which I want to rename to `Circle.expAddChar`, see my previous comments) ?
We also probably want another definition that chooses L for special cases, but we can keep that for future PRs.

RemyDegenne commented:
By the way, I wrote too much of the `probChar` code to merge this myself, so I'll find someone else to do the final review and merge.

RemyDegenne commented on Mathlib/Analysis/Complex/Circle.lean:
```diff
@@ -129,6 +129,25 @@ def expHom : ℝ →+ Additive Circle where
+/-- Exponential map onto the circle, defined as additive character -/
+noncomputable
+def expAddChar : AddChar ℝ Circle where
+  toFun z := Circle.exp (z)
+  map_zero_eq_one' := by simp only; rw [Circle.exp_zero]
+  map_add_eq_mul' x y := by simp only; rw [Circle.exp_add]
```
```suggestion
  toFun := Circle.exp
  map_zero_eq_one' := Circle.exp_zero
  map_add_eq_mul' := Circle.exp_add
```

JakobStiefel commented:
> The name and the placement of the file `FiniteMeasureCharFun` are a bit strange to me. Why is it in the Integral folder? Since it is defined from a measure, I would have put it in the Measure folder. And the name of the file could perhaps be `CharacteristicFunction`?
> 
> You don't actually add a definition for the characteristic function. Perhaps it would be worth adding a definition `charFun P hL w` to refer to the integral of `probChar` for the standard choice of `e = probFourierChar` (which I want to rename to `Circle.expAddChar`, see my previous comments) ? We also probably want another definition that chooses L for special cases, but we can keep that for future PRs.

I added a definition of `charFun P hL w`. I renamed the previous extensionality result and added a new theorem `ext_of_charFun_eq`. The latter is now used to show that `complexMGF` separates measures.

JakobStiefel commented:
> By the way, I wrote too much of the `probChar` code to merge this myself, so I'll find someone else to do the final review and merge.

Alright. My contract ends at the end of march, maybe you can find somebody who is available this week :)

RemyDegenne commented:
@sgouezel could you review this PR? I contributed too much code to it to be comfortable merging myself.

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,249 @@
+    {he : Continuous e} {hL : Continuous fun p : V × W ↦ L p.1 p.2}
+
+/-- The bounded continuous mapping `fun v ↦ e (L v (Multiplicative.toAdd w))` from `V` to `ℂ`. -/
+def probChar (he : Continuous e) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
```
I'm not convinced that the unbundled version should take a parameter in `Multiplicative W`: for most uses, you will apply this to a good old-fashioned element of `W` (especially for the usual characteristic function, which takes a real parameter). So for end users it seems better to take a parameter in `W`. Only when you want to to bundled stuff should you use `Multiplicative W` (and probably the users won't interact as much with these bundled versions). Same thing below.

A good sanity check is to see if you abuse the defeqs or not. Currently, `charFun` is written with a parameter in `W`, but it's an integral of `probChar` taking parameters in `Multiplicative W`, so something fishy is going on...

If I understand correctly, it's only `probCharMonoidHom` that really needs to be defined on `Multiplicative W`, otherwise `W` is fine everywhere.

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,249 @@
+
+- `probChar _ _ w : V →ᵇ ℂ`: The bounded continuous mapping
+  `fun v ↦ e (L v (Multiplicative.toAdd w))` from `V` to `ℂ`, `e` is a continuous additive
+  character and `L : V →ₗ[ℝ] W →ₗ[ℝ] ℝ` is a bilinear map.
```
```suggestion
  character and `L : V →ₗ[ℝ] W →ₗ[ℝ] ℝ` is a continuous bilinear map.
```

sgouezel reviewed (commented):
I'm a little bit uncomfortable with the file organization: there is a whole part of the main file that has nothing to do with measures or characteristic functions, it's rather definitions and theorems around `probChar` (and its algebra versions, and its separating properties). I'd say all this part rather belongs to a new file in `Analysis.Fourier`. And I'm not convinced by the name `probChar` because it's nothing specific to probability theory. Something like `BoundedContinuousFunction.char` would feel better (and by opening `BoundedContinuousFunction` it wouldn't even be too bad to type).

sgouezel commented on Mathlib/Analysis/Fourier/Char.lean:
```diff
@@ -0,0 +1,190 @@
+/-
```
I'd rename the file to BoundedContinuousFunctionChar, because the point of the current file is to work with a character bundled as a continuous bounded function, not a general character.

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,105 @@
+-/
+theorem ext_of_integral_char_eq (he : Continuous e) (he' : e ≠ 1)
+    (hL' : ∀ v ≠ 0, L v ≠ 0) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
+    (P P' : Measure V) [IsFiniteMeasure P] [IsFiniteMeasure P']
```
```suggestion
    {P P' : Measure V} [IsFiniteMeasure P] [IsFiniteMeasure P']
```
They can be inferred from `h` or the conclusion, so better make them implicit.

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,105 @@
+  simp only [smul_eq_mul, MeasureTheory.integral_mul_left, mul_eq_mul_left_iff]
+  exact Or.inl (h i)
+
+lemma expAddChar_ne_one : Circle.expAddChar ≠ 1 := by
```
Better move this just after the definition of `expAddChar`

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,105 @@
+-/
+theorem ext_of_charFun_eq
+    (hL' : ∀ v ≠ 0, L v ≠ 0) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
+    (P P' : Measure V) [IsFiniteMeasure P] [IsFiniteMeasure P'] (h : charFun P hL = charFun P' hL) :
```
```suggestion
    {P P' : Measure V} [IsFiniteMeasure P] [IsFiniteMeasure P'] (h : charFun P hL = charFun P' hL) :
```

sgouezel commented on Mathlib/MeasureTheory/Measure/CharacteristicFunction.lean:
```diff
@@ -0,0 +1,105 @@
+`fun v ↦ e (L v w)` with respect to `P` for the standard choice of `e = Circle.exp`.
+-/
+noncomputable
+def charFun [MeasurableSpace V] (P : Measure V) (hL : Continuous fun p : V × W ↦ L p.1 p.2)
```
What about specializing even further, to an inner product space and `L` the scalar product? For me, this is what the usual characteristic function is, so why not set it up like that for 99% of applications? (but still have your general lemma `ext_of_integral_char_eq` in case someone needs the general situtation)

JakobStiefel replied:
I had a discussion on this with Remy Degenne, we decided that this should be done in a future PR.

In my opinion, the terminology "characteristic function" still makes sense even in a setting without inner product, as it separates finite measures. But I agree that there should be code on the special case `V = W`, where `V` is endowed with an inner product.

hanwenzhu replied:
Just an idea, what about naming `charWithBilinear` and `charFunWithBilinear` for ones with custom `L` and `char` and `charFun` for the default `L` = inner product (in a future PR)?

sgouezel commented on Mathlib/Analysis/Complex/Circle.lean:
```diff
@@ -129,6 +129,27 @@ def expHom : ℝ →+ Additive Circle where
 @[simp] lemma exp_sub (x y : ℝ) : exp (x - y) = exp x / exp y := expHom.map_sub x y
 @[simp] lemma exp_neg (x : ℝ) : exp (-x) = (exp x)⁻¹ := expHom.map_neg x
 
+/-- Exponential map onto the circle, defined as additive character -/
```
There is already `Real.fourierChar`, which is closely related to this one and used throughout Fourier theory (with a specific notation `𝐞`). There is the issue that `Real.fourierChar` uses the analyst convention that there is a 2\pi in the exponential, while here you're using the probabilist convention that there is no 2\pi (and I agree it's the one you want to define the characteristic function). 

What about moving the definition of `Real.fourierChar` (with the two lemmas that follow it and its notation) here, explain that we have two objects because of the two conventions, and try to have a completely symmetric API for one and the other in the current file? 

Then there is the question of naming. `fourierChar` and `expAddChar` don't really fit together. `fourierChar` and `probChar` (as in a previous version of the PR) are maybe better, but I'm not convinced either. Do you have a better idea?

JakobStiefel replied:
Done, both definitions of additive characters are in the current file. 

For now, I use the name `probChar`, as I don't really have a better idea. The names reflect the used convention, (Fourier-) analysis and probability.

hanwenzhu commented on Mathlib/Analysis/Complex/Circle.lean:
```diff
@@ -159,3 +177,49 @@ protected lemma norm_smul {E : Type*} [SeminormedAddCommGroup E] [NormedSpace 
 
 end Circle
+
+open scoped Real
```
```suggestion
```
`Real` doesn't have to be opened since the rest is in `namespace Real`

grunweg commented:
Thanks for your PR. Can you merge the master branch, please? Right now, your PR has a merge conflict, meaning it won't show up in the review queue.

RemyDegenne commented:
The unresolved comments are about names and how much `charFun` should be specialized. Let's try and resolve those issues?
- `probChar` is the current name of the definition which is the same as `fourierChar` up to a 2*pi factor. I think that name is fine, and we don't seem to have a better suggestion.
- The current `charFun` definition chooses that character but does not make a choice of bilinear map. I agree with @sgouezel  that the most standard definition is the particular case of the inner product, and I think we should reserve the name `charFun` for that. I am the one who pushed for the introduction of a characteristic function definition in this PR and suggested the current one, but that was not a great suggestion, sorry. What about we remove the definition and merge the PR without it? That way we can merge all those interesting results, and then we'll add a `charFun` definition later.

sgouezel commented on Mathlib/Analysis/Complex/Circle.lean:
```diff
@@ -159,3 +177,47 @@ protected lemma norm_smul {E : Type*} [SeminormedAddCommGroup E] [NormedSpace 
+
+@[continuity]
+theorem continuous_fourierChar : Continuous 𝐞 := Circle.exp.continuous.comp (continuous_mul_left _)
+
```
In the same way there is `probChar_ne_one`, could you add `fourierChar_ne_one` here, to keep the APIs parallel?

sgouezel reviewed (commented):
I'm fine with removing the `charFun` part and merging the PR, or changing the `charFun` to be defined on an inner product space, as you prefer.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

RemyDegenne commented:
I just pushed 3 commits.
- the first one is a merge of master to resolve a conflict
- the second one adds `fourierChar_ne_one` as requested by Sébastien, and also adds a few other lemmas about `fourierChar` and `probChar`.
- the third one deletes the `charFun` definition.

@JakobStiefel please review those commits and undo anything you don't like. In particular that last commit supposes that we don't define `charFun` in this PR. If you disagree with that, feel free to revert the commit.

hanwenzhu commented on Mathlib/Analysis/Fourier/BoundedContinuousFunctionChar.lean:
```diff
@@ -0,0 +1,190 @@
+    char he hL (-w) = star (char he hL w) := by ext; simp
+
+/-- If `e` and `L` are non-trivial, then `char he hL w, w : W` separates points in `V`. -/
+theorem char_SeparatesPoints (he : Continuous e) (he' : e ≠ 1)
```
I think there is probably a better name that conforms to the naming convention? I would propose `ext_of_char_eq` and also rearrange the signature from `v ≠ v' → ∃ w : W, char he hL w v ≠ char he hL w v'` to `(∀ w : W, char he hL w v = char he hL w v') → v = v'`

JakobStiefel commented:
@RemyDegenne thanks for resolving the issues! The current version of this PR is fine for me. The details of different versions of characteristic functions can be worked out in a future PR
